File: cpp/src/io_device.cpp
/*
    This file is part of the CLI library.  The CLI library aims to provide
    facilities in making text-based user interfaces.
    Copyright (C) 2006-2007 Alexis Royer.

    This library is free software; you can redistribute it and/or
    modify it under the terms of the GNU Lesser General Public
    License as published by the Free Software Foundation; either
    version 2.1 of the License, or (at your option) any later version.

    This library is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
    Lesser General Public License for more details.

    You should have received a copy of the GNU Lesser General Public
    License along with this library; if not, write to the Free Software
    Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
*/


#include <iostream>
#include "cli/io_device.h"
#include "cli/cli.h"
#include "cli/traces.h"

using namespace cli;


const IOEndl cli::endl;

static const TraceClass TRACE_IO_DEVICE_INSTANCES("IO_DEVICE_INSTANCES", "IO device instance management");
static const TraceClass TRACE_IO_DEVICE_OPENING("IO_DEVICE_OPENING", "IO device opening management");


OutputDevice::OutputDevice(
        const std::string& STR_DbgName,
        const std::string& STR_Endl,
        const bool B_AutoDelete)
  : m_strDbgName(STR_DbgName),
    m_strEndl(STR_Endl),
    m_iInstanceLock(B_AutoDelete ? 0 : 1), m_iOpenLock(0)
{
    // Please, no traces in constructor for consistency reasons.
}

OutputDevice::~OutputDevice(void)
{
}

const int OutputDevice::UseInstance(const CallInfo& CLI_CallInfo)
{
    GetTraces().Trace(TRACE_IO_DEVICE_INSTANCES)
        << "One more user for instance " << m_strDbgName << "/" << this << ", "
        << "user count: " << m_iInstanceLock << " -> " << m_iInstanceLock + 1 << ", "
        << "from " << CLI_CallInfo.GetFunction() << " "
        << "at " << CLI_CallInfo.GetFile() << ":" << CLI_CallInfo.GetLine() << endl;
    m_iInstanceLock ++;
    return m_iInstanceLock;
}

const int OutputDevice::FreeInstance(const CallInfo& CLI_CallInfo)
{
    GetTraces().Trace(TRACE_IO_DEVICE_INSTANCES)
        << "One less user for instance " << m_strDbgName << "/" << this << ", "
        << "user count: " << m_iInstanceLock << " -> " << m_iInstanceLock - 1 << ", "
        << "from " << CLI_CallInfo.GetFunction() << " "
        << "at " << CLI_CallInfo.GetFile() << ":" << CLI_CallInfo.GetLine() << endl;
    if (m_iInstanceLock == 1)
    {
        GetTraces().Trace(TRACE_IO_DEVICE_INSTANCES)
            << "Deleting the device " << m_strDbgName << "/" << this << endl;
        delete this;
        return 0;
    }
    else
    {
        m_iInstanceLock --;
        assert(m_iInstanceLock > 0);
        return m_iInstanceLock;
    }
}

const int OutputDevice::GetInstanceUsers(void) const
{
    return m_iInstanceLock;
}

const bool OutputDevice::OpenUp(const CallInfo& CLI_CallInfo)
{
    GetTraces().Trace(TRACE_IO_DEVICE_OPENING)
        << "One more user for instance " << m_strDbgName << "/" << this << ", "
        << "user count: " << m_iOpenLock << " -> " << m_iOpenLock + 1 << ", "
        << "from " << CLI_CallInfo.GetFunction() << " "
        << "at " << CLI_CallInfo.GetFile() << ":" << CLI_CallInfo.GetLine() << endl;

    m_iOpenLock ++;

    if (m_iOpenLock == 1)
    {
        GetTraces().Trace(TRACE_IO_DEVICE_OPENING)
            << "Opening the device " << m_strDbgName << "/" << this << endl;
        if (! OpenDevice())
        {
            return false;
        }
    }

    return true;
}

const bool OutputDevice::CloseDown(const CallInfo& CLI_CallInfo)
{
    bool b_Res = true;

    if (m_iOpenLock > 0)
    {
        GetTraces().Trace(TRACE_IO_DEVICE_OPENING)
            << "One less user for instance " << m_strDbgName << "/" << this << ", "
            << "user count: " << m_iOpenLock << " -> " << m_iOpenLock - 1 << ", "
            << "from " << CLI_CallInfo.GetFunction() << " "
            << "at " << CLI_CallInfo.GetFile() << ":" << CLI_CallInfo.GetLine() << endl;

        if (m_iOpenLock == 1)
        {
            GetTraces().Trace(TRACE_IO_DEVICE_OPENING)
                << "Closing the device " << m_strDbgName << "/" << this << endl;
            b_Res = CloseDevice();
        }

        m_iOpenLock --;
    }
    else
    {
        GetTraces().Trace(TRACE_IO_DEVICE_OPENING)
            << "No more closing down for instance " << m_strDbgName << "/" << this << ", "
            << "user count = " << m_iOpenLock << ", "
            << "from " << CLI_CallInfo.GetFunction() << " "
            << "at " << CLI_CallInfo.GetFile() << ":" << CLI_CallInfo.GetLine() << endl;
    }

    return b_Res;
}

const int OutputDevice::GetOpenUsers(void) const
{
    return m_iOpenLock;
}

const OutputDevice& OutputDevice::operator <<(const std::string& STR_Out) const
{
    PutString(STR_Out);
    return *this;
}

const OutputDevice& OutputDevice::operator <<(const char* const STR_Out) const
{
    if (STR_Out != NULL)
    {
        PutString(STR_Out);
    }
    return *this;
}

const OutputDevice& OutputDevice::operator <<(const unsigned char UC_Out) const
{
    return this->operator<<((unsigned int) UC_Out);
}

const OutputDevice& OutputDevice::operator <<(const char C_Out) const
{
    PutString(std::string() += C_Out);
    return *this;
}

const OutputDevice& OutputDevice::operator <<(const short S_Out) const
{
    return this->operator<<((int) S_Out);
}

const OutputDevice& OutputDevice::operator <<(const unsigned short US_Out) const
{
    return this->operator<<((unsigned int) US_Out);
}

const OutputDevice& OutputDevice::operator <<(const long L_Out) const
{
    return this->operator <<((int) L_Out);
}

const OutputDevice& OutputDevice::operator <<(const unsigned long UL_Out) const
{
    return this->operator <<((unsigned int) UL_Out);
}

const OutputDevice& OutputDevice::operator <<(const int I_Out) const
{
    char str_Out[128];
    snprintf(str_Out, sizeof(str_Out), "%d", I_Out);
    PutString(str_Out);
    return *this;
}

const OutputDevice& OutputDevice::operator <<(const unsigned int UI_Out) const
{
    char str_Out[128];
    snprintf(str_Out, sizeof(str_Out), "%u", UI_Out);
    PutString(str_Out);
    return *this;
}

const OutputDevice& OutputDevice::operator <<(const float F_Out) const
{
    return this->operator <<((double) F_Out);
}

const OutputDevice& OutputDevice::operator <<(const double D_Out) const
{
    char str_Out[128];
    snprintf(str_Out, sizeof(str_Out), "%f", D_Out);
    PutString(str_Out);
    return *this;
}

const OutputDevice& OutputDevice::operator <<(void* const PV_Out) const
{
    char str_Out[128];
    snprintf(str_Out, sizeof(str_Out), "0x%08x", (unsigned int) PV_Out);
    PutString(str_Out);
    return *this;
}

const OutputDevice& OutputDevice::operator <<(const IOEndl& CLI_IOEndl) const
{
    PutString(m_strEndl);
    return *this;
}

OutputDevice& OutputDevice::GetNullDevice(void)
{
    class NullDevice : public OutputDevice
    {
    public:
        NullDevice(void) : OutputDevice("null", "", false) {}
        virtual ~NullDevice(void) {}

    protected:
        virtual const bool OpenDevice(void) { return true; }
        virtual const bool CloseDevice(void) { return true; }
    public:
        virtual void PutString(const std::string& STR_Out) const {}
        virtual void Beep(void) const {}
    };

    static NullDevice cli_Null;
    return cli_Null;
}

OutputDevice& OutputDevice::GetStdOut(void)
{
    class StdOutDevice : public OutputDevice
    {
    public:
        StdOutDevice(void) : OutputDevice("stdout", "\n", false) {}
        virtual ~StdOutDevice(void) {}

    protected:
        virtual const bool OpenDevice(void) { return true; }
        virtual const bool CloseDevice(void) { return true; }
    public:
        virtual void PutString(const std::string& STR_Out) const { std::cout << STR_Out; }
        virtual void Beep(void) const {}
    };

    static StdOutDevice cli_StdOut;
    return cli_StdOut;
}

OutputDevice& OutputDevice::GetStdErr(void)
{
    class StdErrDevice : public OutputDevice
    {
    public:
        StdErrDevice(void) : OutputDevice("stderr", "\n", false) {}
        virtual ~StdErrDevice(void) {}

    protected:
        virtual const bool OpenDevice(void) { return true; }
        virtual const bool CloseDevice(void) { return true; }
    public:
        virtual void PutString(const std::string& STR_Out) const { std::cerr << STR_Out; }
        virtual void Beep(void) const {}
    };

    static StdErrDevice cli_StdErr;
    return cli_StdErr;
}


IODevice::IODevice(
        const std::string& STR_DbgName,
        const std::string& STR_Endl,
        const bool B_AutoDelete)
  : OutputDevice(STR_DbgName, STR_Endl, B_AutoDelete)
{
}

IODevice::~IODevice(void)
{
}

IODevice& IODevice::GetNullDevice(void)
{
    class NullDevice : public IODevice
    {
    public:
        NullDevice(void) : IODevice("null", "", false) {}
        virtual ~NullDevice(void) {}

    protected:
        virtual const bool OpenDevice(void) { return true; }
        virtual const bool CloseDevice(void) { return true; }
    public:
        virtual void PutString(const std::string& STR_Out) const {}
        virtual void Beep(void) const {}
        virtual const KEY GetKey(void) const { return NULL_KEY; }
    };

    static NullDevice cli_Null;
    return cli_Null;
}

IODevice& IODevice::GetStdIn(void)
{
    class StdInDevice : public IODevice
    {
    public:
        StdInDevice(void) : IODevice("stdin", "", false) {
        }
        virtual ~StdInDevice(void) {
        }

    protected:
        virtual const bool OpenDevice(void) {
            OutputDevice::GetStdOut().UseInstance(__CALL_INFO__);
            return OutputDevice::GetStdOut().OpenUp(__CALL_INFO__);
        }
        virtual const bool CloseDevice(void) {
            bool b_Res = OutputDevice::GetStdOut().CloseDown(__CALL_INFO__);
            OutputDevice::GetStdOut().FreeInstance(__CALL_INFO__);
            return b_Res;
        }
    public:
        virtual void PutString(const std::string& STR_Out) const {
            OutputDevice::GetStdOut().PutString(STR_Out);
        }
        virtual void Beep(void) const {
            OutputDevice::GetStdOut().Beep();
        }
        virtual const KEY GetKey(void) const {
            if (! m_strInput.empty())
            {
                const char c_Char = m_strInput[0];
                m_strInput.erase(m_strInput.begin());
                return IODevice::GetKey(c_Char);
            }
            else
            {
                std::cin >> m_strInput;
                m_strInput += "\n";
                return StdInDevice::GetKey();
            }
        }

    private:
        mutable std::string m_strInput;
    };

    static StdInDevice cli_StdIn;
    return cli_StdIn;
}

const KEY IODevice::GetKey(const int I_Char) const
{
    switch (I_Char)
    {
    case 3:             return BREAK;
    case 4:             return LOGOUT;
    case 10: case 13:   return ENTER;
    case 27:            return ESCAPE;
    case 127:           return BACKSPACE;
    case 1001:          return KEY_UP;
    case 1002:          return KEY_DOWN;
    case 1005:          return PAGE_UP;
    case 1006:          return PAGE_DOWN;

    case ' ':           return SPACE;
    case '\t':          return TAB;

    case '0':   return KEY_0;
    case '1':   return KEY_1;
    case '2':   return KEY_2;
    case '3':   return KEY_3;
    case '4':   return KEY_4;
    case '5':   return KEY_5;
    case '6':   return KEY_6;
    case '7':   return KEY_7;
    case '8':   return KEY_8;
    case '9':   return KEY_9;

    case 'a':   return KEY_a;
    case 'b':   return KEY_b;
    case 'c':   return KEY_c;
    case 'd':   return KEY_d;
    case 'e':   return KEY_e;
    case 'f':   return KEY_f;
    case 'g':   return KEY_g;
    case 'h':   return KEY_h;
    case 'i':   return KEY_i;
    case 'j':   return KEY_j;
    case 'k':   return KEY_k;
    case 'l':   return KEY_l;
    case 'm':   return KEY_m;
    case 'n':   return KEY_n;
    case 'o':   return KEY_o;
    case 'p':   return KEY_p;
    case 'q':   return KEY_q;
    case 'r':   return KEY_r;
    case 's':   return KEY_s;
    case 't':   return KEY_t;
    case 'u':   return KEY_u;
    case 'v':   return KEY_v;
    case 'w':   return KEY_w;
    case 'x':   return KEY_x;
    case 'y':   return KEY_y;
    case 'z':   return KEY_z;

    case 'A':   return KEY_A;
    case 'B':   return KEY_B;
    case 'C':   return KEY_C;
    case 'D':   return KEY_D;
    case 'E':   return KEY_E;
    case 'F':   return KEY_F;
    case 'G':   return KEY_G;
    case 'H':   return KEY_H;
    case 'I':   return KEY_I;
    case 'J':   return KEY_J;
    case 'K':   return KEY_K;
    case 'L':   return KEY_L;
    case 'M':   return KEY_M;
    case 'N':   return KEY_N;
    case 'O':   return KEY_O;
    case 'P':   return KEY_P;
    case 'Q':   return KEY_Q;
    case 'R':   return KEY_R;
    case 'S':   return KEY_S;
    case 'T':   return KEY_T;
    case 'U':   return KEY_U;
    case 'V':   return KEY_V;
    case 'W':   return KEY_W;
    case 'X':   return KEY_X;
    case 'Y':   return KEY_Y;
    case 'Z':   return KEY_Z;

    case '+':   return PLUS;
    case '-':   return MINUS;
    case '*':   return STAR;
    case '/':   return SLASH;
    case '<':   return LOWER_THAN;
    case '>':   return GREATER_THAN;
    case '=':   return EQUAL;
    case '%':   return PERCENT;

    case '_':   return UNDERSCORE;
    case '@':   return AROBASE;
    case '#':   return SHARP;
    case '&':   return AMPERCENT;
    case '$':   return DOLLAR;
    case '\\':  return BACKSLASH;
    case '|':   return PIPE;

    case '?':   return QUESTION;
    case '!':   return EXCLAMATION;
    case ':':   return COLUMN;
    case '.':   return DOT;
    case ',':   return COMA;
    case ';':   return SEMI_COLUMN;
    case '\'':  return QUOTE;
    case '"':   return DOUBLE_QUOTE;

    case '(':   return OPENING_BRACE;
    case ')':   return CLOSING_BRACE;
    case '{':   return OPENING_CURLY_BRACE;
    case '}':   return CLOSING_CURLY_BRACE;
    case '[':   return OPENING_BRACKET;
    case ']':   return CLOSING_BRACKET;

    default:
        // Unrecognized character.
        return NULL_KEY;
    }
}

File: java/src/cli/IODevice.java
/*
    Copyright (c) 2006-2018, Alexis Royer, http://alexis.royer.free.fr/CLI

    All rights reserved.

    Redistribution and use in source and binary forms, with or without modification, are permitted provided that the following conditions are met:

        * Redistributions of source code must retain the above copyright notice, this list of conditions and the following disclaimer.
        * Redistributions in binary form must reproduce the above copyright notice, this list of conditions and the following disclaimer in the documentation
          and/or other materials provided with the distribution.
        * Neither the name of the CLI library project nor the names of its contributors may be used to endorse or promote products derived from this software
          without specific prior written permission.

    THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
    "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
    LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
    A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR
    CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
    EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
    PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
    PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
    LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
    NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
    SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
*/

package cli;


/** Input/output device management.

    See OutputDevice comment. */
public abstract class IODevice {

    /** Generic input/output device interface. */
    public interface Interface extends OutputDevice.Interface {
        /** Input key capture.
            @return Key captured. */
        public int getKey();

        /** Location accessor.
            @return The location as a resource string. */
        public ResourceString getLocation();

        /** Stack overflow protection.
            @param CLI_Device Other device that the callee device should check it would input characters from.
            @return true if the callee device would redirect input to the given device for reading.

            Determines whether the current device would input the given device in any ways.
            Default implementation checks whether CLI_Device is the self device. */
        public boolean wouldInput(IODevice.Interface CLI_Device);
    }

    /** Native-implemented input/output devices. */
    public static abstract class Native extends OutputDevice.Native implements IODevice.Interface {
        /** Constructor.
            @param I64_NativeRef Native instance reference. */
        protected Native(long I64_NativeRef) {
            super(I64_NativeRef);
        }

        // IODevice.Interface native input/output device implementation.

        public final int getKey() {
            return IODevice.__Native__getKey(this.getNativeRef());
        }

        public ResourceString getLocation() {
            ResourceString cli_Location = (ResourceString) NativeObject.getObject(IODevice.__Native__getLocation(this.getNativeRef()));
            NativeObject.forget(cli_Location);
            return cli_Location;
        }

        public boolean wouldInput(IODevice.Interface CLI_Device) {
            return IODevice.__Native__wouldInput(this.getNativeRef(), CLI_Device.getNativeRef());
        }
    }
    // JNI seems to have trouble at linking following methods when they are embedded in the nested Native class above (at least with java 1.5.0_03).
    // Therefore they are just declared in the scope of the global IODevice class with a __Native prefix.
    private static final native int __Native__getKey(long I64_NativeDeviceRef);
    private static final native int __Native__getLocation(long I64_NativeDeviceRef);
    private static final native boolean __Native__wouldInput(long I64_NativeIODeviceRef, long I64_NativeIODevice2Ref);

    /** Java-implemented input/output devices. */
    public static abstract class Java extends OutputDevice.Java implements IODevice.Interface {
        /** Constructor.
            @param J_DbgName Debug name. Useful for traces only. */
        protected Java(String J_DbgName) {
            super(IODevice.__Java__Java(J_DbgName));
        }

        /** Constructor for NonBlockingIODevice classes only.
            @param I64_NativeRef Native instance reference. */
        protected Java(long I64_NativeRef) {
            super(I64_NativeRef);
        }

        // IODevice.Interface Java input/output device implementation.

        public abstract int getKey();
        private final int __getKey() {
            Traces.safeTrace(NativeTraces.CLASS, this, NativeTraces.begin("IODevice.Java.__getKey()"));
            int e_Key = getKey();
            Traces.safeTrace(NativeTraces.CLASS, this, NativeTraces.end("IODevice.Java.__getKey()", new Integer(e_Key).toString()));
            return e_Key;
        }

        public abstract ResourceString getLocation();
        private final void __getLocation(long I64_NativeResourceStringRef) {
            Traces.safeTrace(NativeTraces.CLASS, this, NativeTraces.begin("IODevice.Java.__getLocation(I64_NativeResourceStringRef)"));
            Traces.safeTrace(NativeTraces.CLASS, this, NativeTraces.param("I64_NativeResourceStringRef", new Long(I64_NativeResourceStringRef).toString()));
            // Retrieve the native resource string instance (output parameter).
            ResourceString cli_NativeLocation = (ResourceString) NativeObject.getObject(I64_NativeResourceStringRef);
            if (cli_NativeLocation != null) {
                // Make the call to the Java handler.
                ResourceString cli_JavaLocation = getLocation();
                if (cli_JavaLocation != null) {
                    // Report information from Java to native instance.
                    for (int e_Lang = 0; e_Lang < ResourceString.LANG_COUNT; e_Lang++) {
                        Traces.safeTrace(NativeTraces.CLASS, this, NativeTraces.value("e_Lang", new Integer(e_Lang).toString()));
                        String j_Location = cli_JavaLocation.getString(e_Lang);
                        Traces.safeTrace(NativeTraces.CLASS, this, NativeTraces.value("j_Location", j_Location));
                        cli_NativeLocation.setString(e_Lang, j_Location);
                    }
                }
            }
            Traces.safeTrace(NativeTraces.CLASS, this, NativeTraces.end("IODevice.Java.__getLocation()"));
        }

        public boolean wouldInput(IODevice.Interface CLI_Device) {
            return (CLI_Device == this);
        }
        private final boolean __wouldInput(long I64_NativeDeviceRef) {
            // Do not trace! for consistency reasons.
            //Traces.safeTrace(NativeTraces.CLASS, this, NativeTraces.begin("OutputDevice.Java.__wouldOutput(CLI_Device)"));
            boolean b_Res = false;
            try {
                IODevice.Interface cli_Device = (IODevice.Interface) NativeObject.getObject(I64_NativeDeviceRef);
                if (cli_Device != null) {
                    b_Res = wouldInput(cli_Device);
                }
            } catch (Exception e) {
            }
            //Traces.safeTrace(NativeTraces.CLASS, this, NativeTraces.end("OutputDevice.Java.__wouldOutput()"));
            return b_Res;
        }
    }
    // JNI seems to have trouble at linking following methods when they are embedded in the nested Java class above (at least with java 1.5.0_03).
    // Therefore they are just declared in the scope of the global IODevice class with a __Java prefix.
    private static final native long __Java__Java(String J_DbgName);


    // General static input/output device features.

    /** Null device singleton.
        @return The null output device. */
    public static final IODevice.Interface getNullDevice() {
        class NullDevice extends IODevice.Native {
            public NullDevice() {
                super(__getNullDevice());
            }
        }
        if (m_cliNullDevice == null) {
            m_cliNullDevice = new NullDevice();
        }
        return m_cliNullDevice;
    }
    private static final native long __getNullDevice();
    private static IODevice.Interface m_cliNullDevice = null;

    /** Standard input device singleton.
        @return The standard input device. */
    public static final IODevice.Interface getStdIn() {
        class StdInDevice extends IODevice.Native {
            public StdInDevice() {
                super(__getStdIn());
            }
        }
        if (m_cliStdIn == null) {
            m_cliStdIn = new StdInDevice();
        }
        return m_cliStdIn;
    }
    private static final native long __getStdIn();
    private static IODevice.Interface m_cliStdIn = null;

    /** @brief Common char translation.
        @param I_Char Common char to translate.
        @return KEY code corresponding to the given common char.

        No JNI call to native implementation, but full Java re-implementation in order to avoid strange Java/native character translations. */
    static final int getKey(int I_Char) {
        switch (I_Char) {
        case 10: case 13:   return OutputDevice.ENTER;

        case ' ':   return OutputDevice.SPACE;
        case '\t':  return OutputDevice.TAB;
        case '\b':  return OutputDevice.BACKSPACE;

        case '0':   return OutputDevice.KEY_0;
        case '1':   return OutputDevice.KEY_1;
        case '2':   return OutputDevice.KEY_2;
        case '3':   return OutputDevice.KEY_3;
        case '4':   return OutputDevice.KEY_4;
        case '5':   return OutputDevice.KEY_5;
        case '6':   return OutputDevice.KEY_6;
        case '7':   return OutputDevice.KEY_7;
        case '8':   return OutputDevice.KEY_8;
        case '9':   return OutputDevice.KEY_9;

        case 'a':   return OutputDevice.KEY_a;
        case 0xe1:  // iso-8859-1 encoding for 'á'
                    return OutputDevice.KEY_aacute;
        case 0xe0:  // iso-8859-1 encoding for 'à'
                    return OutputDevice.KEY_agrave;
        case 0xe4:  // iso-8859-1 encoding for 'ä'
                    return OutputDevice.KEY_auml;
        case 0xe2:  // iso-8859-1 encoding for 'â'
                    return OutputDevice.KEY_acirc;
        case 'b':   return OutputDevice.KEY_b;
        case 'c':   return OutputDevice.KEY_c;
        case 0xe7:  // iso-8859-1 encoding for 'ç'
                    return OutputDevice.KEY_ccedil;
        case 'd':   return OutputDevice.KEY_d;
        case 'e':   return OutputDevice.KEY_e;
        case 0xe9:  // iso-8859-1 encoding for 'é'
                    return OutputDevice.KEY_eacute;
        case 0xe8:  // iso-8859-1 encoding for 'è'
                    return OutputDevice.KEY_egrave;
        case 0xeb:  // iso-8859-1 encoding for 'ë'
                    return OutputDevice.KEY_euml;
        case 0xea:  // iso-8859-1 encoding for 'ê'
                    return OutputDevice.KEY_ecirc;
        case 'f':   return OutputDevice.KEY_f;
        case 'g':   return OutputDevice.KEY_g;
        case 'h':   return OutputDevice.KEY_h;
        case 'i':   return OutputDevice.KEY_i;
        case 0xed:  // iso-8859-1 encoding for 'í'
                    return OutputDevice.KEY_iacute;
        case 0xec:  // iso-8859-1 encoding for 'ì'
                    return OutputDevice.KEY_igrave;
        case 0xef:  // iso-8859-1 encoding for 'ï'
                    return OutputDevice.KEY_iuml;
        case 0xee:  // iso-8859-1 encoding for 'î'
                    return OutputDevice.KEY_icirc;
        case 'j':   return OutputDevice.KEY_j;
        case 'k':   return OutputDevice.KEY_k;
        case 'l':   return OutputDevice.KEY_l;
        case 'm':   return OutputDevice.KEY_m;
        case 'n':   return OutputDevice.KEY_n;
        case 'o':   return OutputDevice.KEY_o;
        case 0xf3:  // iso-8859-1 encoding for 'ó'
                    return OutputDevice.KEY_oacute;
        case 0xf2:  // iso-8859-1 encoding for 'ò'
                    return OutputDevice.KEY_ograve;
        case 0xf6:  // iso-8859-1 encoding for 'ö'
                    return OutputDevice.KEY_ouml;
        case 0xf4:  // iso-8859-1 encoding for 'ô'
                    return OutputDevice.KEY_ocirc;
        case 'p':   return OutputDevice.KEY_p;
        case 'q':   return OutputDevice.KEY_q;
        case 'r':   return OutputDevice.KEY_r;
        case 's':   return OutputDevice.KEY_s;
        case 't':   return OutputDevice.KEY_t;
        case 'u':   return OutputDevice.KEY_u;
        case 0xfa:  // iso-8859-1 encoding for 'ú'
                    return OutputDevice.KEY_uacute;
        case 0xf9:  // iso-8859-1 encoding for 'ù'
                    return OutputDevice.KEY_ugrave;
        case 0xfc:  // iso-8859-1 encoding for 'ü'
                    return OutputDevice.KEY_uuml;
        case 0xfb:  // iso-8859-1 encoding for 'û'
                    return OutputDevice.KEY_ucirc;
        case 'v':   return OutputDevice.KEY_v;
        case 'w':   return OutputDevice.KEY_w;
        case 'x':   return OutputDevice.KEY_x;
        case 'y':   return OutputDevice.KEY_y;
        case 'z':   return OutputDevice.KEY_z;

        case 'A':   return OutputDevice.KEY_A;
        case 'B':   return OutputDevice.KEY_B;
        case 'C':   return OutputDevice.KEY_C;
        case 'D':   return OutputDevice.KEY_D;
        case 'E':   return OutputDevice.KEY_E;
        case 'F':   return OutputDevice.KEY_F;
        case 'G':   return OutputDevice.KEY_G;
        case 'H':   return OutputDevice.KEY_H;
        case 'I':   return OutputDevice.KEY_I;
        case 'J':   return OutputDevice.KEY_J;
        case 'K':   return OutputDevice.KEY_K;
        case 'L':   return OutputDevice.KEY_L;
        case 'M':   return OutputDevice.KEY_M;
        case 'N':   return OutputDevice.KEY_N;
        case 'O':   return OutputDevice.KEY_O;
        case 'P':   return OutputDevice.KEY_P;
        case 'Q':   return OutputDevice.KEY_Q;
        case 'R':   return OutputDevice.KEY_R;
        case 'S':   return OutputDevice.KEY_S;
        case 'T':   return OutputDevice.KEY_T;
        case 'U':   return OutputDevice.KEY_U;
        case 'V':   return OutputDevice.KEY_V;
        case 'W':   return OutputDevice.KEY_W;
        case 'X':   return OutputDevice.KEY_X;
        case 'Y':   return OutputDevice.KEY_Y;
        case 'Z':   return OutputDevice.KEY_Z;

        case '+':   return OutputDevice.PLUS;
        case '-':   return OutputDevice.MINUS;
        case '*':   return OutputDevice.STAR;
        case '/':   return OutputDevice.SLASH;
        case '<':   return OutputDevice.LOWER_THAN;
        case '>':   return OutputDevice.GREATER_THAN;
        case '=':   return OutputDevice.EQUAL;
        case '%':   return OutputDevice.PERCENT;

        case '_':   return OutputDevice.UNDERSCORE;
        case '@':   return OutputDevice.AROBASE;
        case '#':   return OutputDevice.SHARP;
        case '&':   return OutputDevice.AMPERCENT;
        case '$':   return OutputDevice.DOLLAR;
        case '\\':  return OutputDevice.BACKSLASH;
        case '|':   return OutputDevice.PIPE;
        case '~':   return OutputDevice.TILDE;
        case 0xb2:  // iso-8859-1 encoding for '²'
                    return OutputDevice.SQUARE;
        case 0x80:  // iso-8859-1 encoding for '€'
                    return OutputDevice.EURO;
        case 0xa3:  // iso-8859-1 encoding for '£'
                    return OutputDevice.POUND;
        case 0xb5:  // iso-8859-1 encoding for 'µ'
                    return OutputDevice.MICRO;
        case 0xa7:  // iso-8859-1 encoding for '§'
                    return OutputDevice.PARAGRAPH;
        case 0xb0:  // iso-8859-1 encoding for '°'
                    return OutputDevice.DEGREE;
        case 0xa9:  // iso-8859-1 encoding for '©'
                    return OutputDevice.COPYRIGHT;

        case '?':   return OutputDevice.QUESTION;
        case '!':   return OutputDevice.EXCLAMATION;
        case ':':   return OutputDevice.COLUMN;
        case '.':   return OutputDevice.DOT;
        case ',':   return OutputDevice.COMA;
        case ';':   return OutputDevice.SEMI_COLUMN;
        case '\'':  return OutputDevice.QUOTE;
        case '"':   return OutputDevice.DOUBLE_QUOTE;

        case '(':   return OutputDevice.OPENING_BRACE;
        case ')':   return OutputDevice.CLOSING_BRACE;
        case '{':   return OutputDevice.OPENING_CURLY_BRACE;
        case '}':   return OutputDevice.CLOSING_CURLY_BRACE;
        case '[':   return OutputDevice.OPENING_BRACKET;
        case ']':   return OutputDevice.CLOSING_BRACKET;

        default:
            // Unrecognized character.
            return OutputDevice.NULL_KEY;
        }
    }

}
